apps/saleor-app-checkout/backend/payments/providers/adyen/middlewares.ts

```diff
@@ -0,0 +1,114 @@
+import { getPrivateSettings } from "@/saleor-app-checkout/backend/configuration/settings";
+import { envVars } from "@/saleor-app-checkout/constants";
+import { PrivateSettingsValues } from "@/saleor-app-checkout/types";
+import { unpackPromise } from "@/saleor-app-checkout/utils/promises";
+import { Types } from "@adyen/api-library";
+import type { Middleware } from "retes";
+import { Response } from "retes/response";
+import { verifyBasicAuth } from "./utils";
+import { adyenHmacValidator, validateHmac } from "./validator";
+
+export type AdyenRequestContext = Required<
+  PrivateSettingsValues<"unencrypted">[keyof PrivateSettingsValues<"unencrypted">]["adyen"]
+>;
+
+export type AdyenRequestParams = Types.notification.Notification;
+
+export const withAdyenWebhookCredentials: Middleware =
+  (handler) => async (request) => {
+    const [error, settings] = await unpackPromise(
+      getPrivateSettings(envVars.apiUrl, false)
+    );
+
+    if (error) {
+      return Response.InternalServerError(
+        "Cannot fetch Adyen API configuration"
+      );
+    }
+
+    const {
+      paymentProviders: { adyen },
+    } = settings;
+
+    const keys = new Set(Object.keys(adyen));
+    const expectedKeys: Array<keyof AdyenRequestContext> = [
```

Comment:
If we ever add anything to `AdyenProviderSettingID` this code should fail – but it won't.

We should modify how `AdyenProviderSettingID` is created and instead infer it from a tuple like so:

```ts
export const adyenProviderSettingIDs = [
  "hmac",
  "apiKey",
  "username",
  "password",
  "clientKey",
  "merchantAccount",
] as const;
export type AdyenProviderSettingID = typeof adyenProviderSettingIDs[number];
```

Reply:
That's perfect idea 🙌🏻 
I totally forgot about this type, but it totally makes sense to use it here